rfcs/subsuites.md

```diff
@@ -0,0 +1,348 @@
+will enforce the property that the subsuite name corresponds to a
+subsuite that has previously been added with `add_subsuite`.
+
+### wptrunner changes
+
+wptrunner already has the concept of test groups. These are sets of
+tests which run in a single browser process, between which the browser
+is restarted. Parallel test running is enabled by scheduling multiple
+test groups, each of which is associated with a single browser
+instance.
+
+It also has the concept of multiple test types (e.g. testharness,
+reftest), each of which has a different implementation. Each test
+group is associated with a single type of test (i.e. a single group
+can contain either testharness tests or reftests, but not both)
+i.e. we can form a mapping from a test type to a list of test
+groups. Groups of different test types can be run in parallel.
+
+For subsuites the different browser configurations in each subsuite
+mean that it's generally desirable not to reuse browser instances
+between different subsuites either by necessity (e.g. because the
+configuration of the browser for the subsuite has to happen at browser
+startup) or because of isolation (so that we don't accidentally get
+shared browser state between subsuites affecting the results of
+later-run tests). Therefore we can assume that each test group should
+be associated with exactly one subsuite. This leaves a small amount of
+parallelism on the table in cases where it would be possible to run
+multiple subsuites in the same group, but the resulting model is
+relatively easy to understand and satisfies the main use cases where
+changing configuration at runtime is not possible.
+
+Therefore the implementation will work as follows:
+
+* Test groups are associated with a pair `(subsuite, test_type)`
+  rather than a test type alone.
+
+* Each `(subsuite, test_type)` combination has a unique
+  `TestImplementation`, which corresponds to a unique instance of the
+  product-specific `Browser` and `Executor` classes.
+
+* Each product implementation is responsible for configuring the
+  browser and executor classes as required for the subsuite. To
+  facilitate this the subsuite details are passed to the product
+  specific functions for getting the browser and executor kwargs.
+
+* Any global state (e.g. reftest screenshot cache) that depends on the
+  test id is instead keyed on `(subsuite, test.id)`.
+
+#### Subsuite definition
+
+The definition of subsuites contains the following properties:
+
+* `name` (required) - The name of the subsuite.
+
+* `config` (required) - Product-specific configuration options for the
+  subsuite. This is used to e.g. set the command line arguments used
+  when launching the browser, or the prefs set in the browser profile.
+
+* `run_info` (optional) - Any updates to the `run_info` associated with the
+  subsuite, other than its name (the `run_info["subsuite"] = name` key
+  is set implicitly).
+
+* `include` (optional) - A list of test id prefixes to include when
```

Comment:
jgraham@, I just see this is optional, which means the subsuite includes all tests?